### PR TITLE
Speed Up Schedule Loading

### PR DESCRIPTION
**Pull request by yaxollum**
Speed up the loading of the schedule on the home page by embedding the schedule data in `index.html`
- [x] Users who are logged in and have added their timetable should see their specific schedule
- [x] Users who are logged in but haven't added their timetable should see a generic schedule and also a message nudging them to add their schedule (#112)
- [x] Users who aren't logged in should see a generic schedule (`scheduleIsPersonal` should be `false`)
- [x] Schedule should work offline for all users
- [x] Links in nudge message should be clickable
- [x] Test regular stuff
- [x] Test winter break (schedules should still count as personal)

### CODE REVIEW

File: core/templates/core/snippets/banner.html [hunk 1]
```diff
@@ -2,17 +2,10 @@
 {% load settings_tags %}
 <div class="banner">
     <div class="background"><img src="{% settings_value 'THEME_BANNER' %}"/></div>
-    {% if banner_message %}
-        <div class="overlay-container">
-            <div class="banner-message center-align">
-                <span>{{ banner_message }}</span>
-            </div>
-        </div>
-    {% endif %}
```
Comment (colourdelete) [inline core/templates/core/snippets/banner.html:11]: Why is this moved down? (BTW this comment isn't important at all, just curious)

File: core/utils/get_schedule.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    date = timezone.localdate()
+    current_term = models.Term.get_current(target_date=date)
+    if current_term is None:
+        return ""
+    elif not info.logged_in:
+        sign_up_url = reverse("account_signup")
+        return f"<a href='{sign_up_url}'>Sign up</a> and add your timetable to see a personalized schedule here."
+    elif info.nudge_add_timetable:
+        add_timetable_url = reverse("timetable_create", args=[current_term.id])
+        return f"<a href='{add_timetable_url}'>Add your timetable</a> to see a personalized schedule here."
```
Comment (colourdelete) [inline core/utils/get_schedule.py:92]: Perhaps it's better to put this in the template, not here? (It's not terribly clear at first glance where this goes to) (I can make the change if you're fine with it)

File: core/utils/get_schedule.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    # generic day schedule is personal if it is empty
+    is_personal = len(schedule) == 0
+    return DaySchedule(schedule, is_personal)
+
+
+def get_week_schedule(user) -> dict:
+    date = timezone.localdate()
+
+    if user.is_authenticated:
+        result = {}  # TODO: use a dictionary comprehension
```
Comment (colourdelete) [inline core/utils/get_schedule.py:47]: A dictionary comprehension may force the code to be less readable for something this complex. I'd suggest not making it a dictionary comprehension. (Also if possible can you make it so the loop bodies are referentially transparent (i.e. not change `date`)?) (Also line 50 may be an unnecessary call in some cases)

An example implementation (sorry it's probably not formatted correctly):
```python3
result = {
    target_date.isoformat():
        (day_schedule := DaySchedule(user.schedule(target_date=date), True)) # first try using personal day schedule
            if len(day_schedule.schedule) != 0 else
            generic_day_schedule(models.Term.get_current(target_date=date), date) # switch to generic day schedule if personal day schedule is empty
    for target_date in [
        date + datetime.timedelta(days) for days in range(7)
    ]
}
```

**Review by colourdelete — APPROVED** [review]
LGTM (except the few minor issues, but those aren't that important)

File: core/templates/core/snippets/banner.html (hunk 1 above)
Comment (yaxollum) [reply core/templates/core/snippets/banner.html:11]: It's because in the original position, you can't click any of the links in the banner message when it the other containers overlay over it.

File: core/utils/get_schedule.py (hunk 2 above)
Comment (yaxollum) [reply core/utils/get_schedule.py:92]: > It's not terribly clear at first glance where this goes to

Yeah, it probably better to put it into the template.

> I can make the change if you're fine with it

That would be great, thanks!

File: core/utils/get_schedule.py (hunk 3 above)
Comment (yaxollum) [reply core/utils/get_schedule.py:47]: I was thinking we could move the loop body into a separate function.

File: core/utils/get_schedule.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    return WeekScheduleInfo(
+        json_data=json.dumps(data, cls=JSONEncoder),
+        nudge_add_timetable=not all(
+            day_schedule.is_personal for day_schedule in data.values()
+        ),
+        logged_in=user.is_authenticated,
+    )
+
+
+def get_schedule_nudge_message(info: WeekScheduleInfo) -> str:
```
[the remark this answers is not in the document]
Comment (yaxollum) [reply core/utils/get_schedule.py:82]: Great idea, thanks! And yes, I'm fine with it :)

File: core/utils/get_schedule.py (hunk 3 above)
Comment (yaxollum) [reply core/utils/get_schedule.py:47]: Something like
```python
def get_day_schedule(user,date) -> DaySchedule:
# implementation here
```

Comment (yaxollum) [reply core/utils/get_schedule.py:47]: (and then use `get_day_schedule` in the dictionary comprehension)

Comment (yaxollum) [reply core/utils/get_schedule.py:47]: > Also line 50 may be an unnecessary call in some cases

Yeah, but I couldn't think of a simple way to avoid it.

Comment (colourdelete) [reply core/utils/get_schedule.py:47]: I'll do the stuff in around an hour.

**Review by colourdelete — APPROVED** [review]

**Comment by yaxollum** [comment]
@colourdelete thanks for all your help on this!

File: core/templates/core/index.html [hunk 5]
```diff
@@ -114,5 +114,5 @@ <h5 class="title truncate">{{ announcement.title }}</h5>
     <meta property="og:url" content="/">
     <meta name="og:image" content="/static/core/img/logo/logo-any-192.png">
     <meta name="theme-color" content="#073763">
-    <script type="text/javascript">{{ schedule_data_js }}</script>
+    <script type="text/javascript">let index_page_data={{ banner_data.json_data }}</script>
```
Comment (colourdelete) [inline core/templates/core/index.html:117]: Whoops thanks about that!

**Review by colourdelete — APPROVED** [review]
LGTM
